File: test/csharpxml/Test1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Runtime.Serialization; // DataContractSerializer (which we may or may not use)
using System.Xml.Serialization;
//using System.ServiceModel;

namespace csharpxml.Test1
{
    // NOTE: if enum does not handle zero and enum prop is not given initial non-zero value, serializer throws.
    public enum MyEnum { zero = 0, one = 1, two = 2, three = 3 };
    [DataContract]
    //[XmlSerializerFormat] // XML attributes in DataContract is largely a hack
    public class TestClass : XmlTestObject<TestClass>
    {
        //[DataMember, XmlAttribute]
        //public string SomeAttr;
        [DataMember]
        public MyEnum MyEnumProp;
        [DataMember]
        public MyEnum[] MyEnumArray;
        [DataMember]
        public bool MyBool;
        [DataMember]
        public byte MyByte;
        [DataMember]
        public int MyInt;
        [DataMember]
        public uint MyUInt;
        [DataMember]
        public long MyLong;
        [DataMember]
        public double MyDouble;
        [DataMember]
        public int? MyNullInt;
        [DataMember]
        public int?[] MyNullIntArr;
        [DataMember]
        public int[] MyIntArray;
        [DataMember]
        public int[][] MyJagIntArray;
        [DataMember]
        public MyEnum[][] MyJagEnumArray;

        [DataMember]
        public SubTestClass MySubClass = new SubTestClass();
        [DataMember]
        public SubTestClass[] MySubClassArray;
        [DataMember]
        public Dictionary<string, int> MyIntDict = new Dictionary<string, int>();
        [DataMember]
        public Dictionary<string, int>[] MyIntDictArr;
        [DataMember]
        public Dictionary<string, SubTestClass> MySubClassDict = new Dictionary<string, SubTestClass>();
        [DataMember]
        public Dictionary<string, MyEnum[]> MyEnumArrDict;
        [DataMember]
        public Dictionary<string, Dictionary<string, int>> MyIntDictDict;
        [DataMember]
        public Dictionary<string, Dictionary<string, int>[]> MyIntDictArrDict;

        public override void PopulateTestObject(int subtest)
        {
            switch (subtest)
            {
                case 0:
                    // run with constructor defaults
                    break;
                case 1:
                    //SomeAttr = "att";
                    MyEnumProp = MyEnum.three;
                    MyEnumArray = new MyEnum[] { MyEnum.one, MyEnum.two };
                    MyBool = true;
                    MyInt = 1;
                    MyNullInt = 42;
                    MyNullIntArr = new int?[] { 43, null, 45 };
                    MyByte = byte.MaxValue;
                    MyUInt = uint.MaxValue;
                    MyLong = long.MaxValue;
                    MyDouble = double.MaxValue;
                    MyIntArray = new int[] { 3, 4, 5 };
                    MyJagIntArray = new int[][]
                    {
                        new int[] { 6, 7, 8 },
                        new int[] { 9, 10 },
                    };
                    MyJagEnumArray = new MyEnum[][]
                    {
                        new MyEnum[] { MyEnum.one, MyEnum.two},
                        new MyEnum[] { MyEnum.three }
                    };
                    MySubClass.Populate(-1);
                    MySubClassArray = new SubTestClass[2];
                    for (int i = 0; i < MySubClassArray.Length; ++i)
                    {
                        MySubClassArray[i] = new SubTestClass();
                        MySubClassArray[i].Populate(i);
                    }
                    MyIntDict.Add("first", 55);
                    MyIntDict.Add("second", 65);
                    MyIntDictArr = new Dictionary<string, int>[2];
                    for (int i = 0; i < MyIntDictArr.Length; ++i)
                    {
                        MyIntDictArr[i] = new Dictionary<string, int>();
                        MyIntDictArr[i].Add("item1-" + i.ToString(), i * 9);
                        MyIntDictArr[i].Add("item2-" + i.ToString(), i * 19);
                    }
                    var sub = new SubTestClass();
                    sub.Populate(11);
                    MySubClassDict.Add("subclass", sub);
                    MyEnumArrDict = new Dictionary<string, MyEnum[]>();
                    MyEnumArrDict.Add("EnumKey", new MyEnum[] { MyEnum.three, MyEnum.two, MyEnum.one });
                    // good lord
                    MyIntDictDict = new Dictionary<string, Dictionary<string, int>>();
                    MyIntDictDict.Add("really", MyIntDict);
                    // stop, please
                    MyIntDictArrDict = new Dictionary<string, Dictionary<string, int>[]>();
                    MyIntDictArrDict.Add("yes", MyIntDictArr);
                    break;
                default:
                    break;
            }
        }
    } // END CLASS: TestClass

    [DataContract]
    public class SubTestClass
    {
        /*[DataMember, XmlAttribute]
        public int SubAttr;
        [DataMember, XmlAttribute]
        public string SubAttr2;*/
        [DataMember]
        public int MySubInt;
        [DataMember]
        public string MySubStr;

        public void Populate(int test)
        {
            //SubAttr = test * 9;
            //SubAttr2 = "att_" + test.ToString();
            //SubAttr3 = test * 42;
            MySubInt = test * 30 + 15;
            MySubStr = "SubClass string " + test.ToString();
        }
    } // END CLASS: SubTestClass

} // END NAMESPACE

File: test/Test1.js
'use strict'
const xml = require('../xml-csharp-cereal');

// DataContract names get messy.
// Some names get little hash suffixes

module.exports.GetFactory = function()
{
    return (new xml.XmlTemplateFactory(TestClass,SubTestClass))
        .addEnum('MyEnum', MyEnumSimple)
        .addDict('ArrayOfKeyValueOfstringint','KeyValueOfstringint',['Key','string'],['Value','int'])
        .addDict('ArrayOfArrayOfKeyValueOfstringint','KeyValueOfstringArrayOfArrayOfKeyValueOfstringintty7Ep6D1',['Key','string'],['Value','ArrayOfKeyValueOfstringint',1])
        .addDict('ArrayOfKeyValueOfstringSubTestClassjHm_ShSx9','KeyValueOfstringSubTestClassjHm_ShSx9',['Key','string'],['Value','SubTestClass'])
        .addDict('ArrayOfKeyValueOfstringArrayOfMyEnumjHm_ShSx9','KeyValueOfstringArrayOfMyEnumjHm_ShSx9', ['Key','string'], ['Value', 'MyEnum', 1])
        .addDict('ArrayOfKeyValueOfstringArrayOfKeyValueOfstringintty7Ep6D1','KeyValueOfstringArrayOfKeyValueOfstringintty7Ep6D1',['Key','string'],['Value','ArrayOfKeyValueOfstringint'])
        .applyDataContractNameSpaces("http://schemas.datacontract.org/2004/07/csharpxml.Test1");
}

class TestClass
{
    constructor()
    {
        // This constructor should mirror its C# counterpart, otherwise missing tags might result in differences due to mismatched defaults.
        // If the C# constructor leaves something null, also null it here.
        // If the C# constructor inits a member, do so here.
        //this.SomeAttr = null; // [XmlAttribute] public string SomeAttr;
        this.MyEnumProp = 0; //public MyEnum MyEnumProp; // C# inits enums as zero
        this.MyEnumArray=null; //public MyEnum[] MyEnumArray;
        this.MyBool=false;
        this.MyByte=0; //public byte MyByte;
        this.MyInt=0; //public int MyInt;
        this.MyUInt=0; //public uint MyUInt;
        this.MyLong=0; //public long MyLong;
        this.MyDouble=0; //public double MyDouble;
        this.MyNullInt; //public int? MyNullInt;
        this.MyNullIntArr; //public int?[] MyNullIntArr;
        this.MyIntArray=null; //public int[] MyIntArray;
        this.MyJagIntArray=null; //public int[][] MyJagIntArray;
        this.MyJagIntArray2=null; //public int[][][] MyJagIntArray2;
        this.MyJagEnumArray=null; //public MyEnum[][] MyJagEnumArray;
        this.MyStrList=null; //public List<string> MyStrList;

        this.MySubClass=new SubTestClass(); //public SubTestClass MySubClass
        this.MySubClassArray=null; //public SubTestClass[] MySubClassArray;
        this.MyIntDict=[]; //public SerializableDictionary<string, int> MyIntDict
        this.MyIntDictArr=null; //public SerializableDictionary<string, int>[] MyIntDictArr;
        this.MySubClassDict=[]; //public SerializableDictionary<string, SubTestClass> MySubClassDict
        this.MyEnumArrDict = null; //public SerializableDictionary<string, MyEnum[]> MyEnumArrDict;
        this.MyIntDictDict = null; //public SerializableDictionary<string, SerializableDictionary<string, int> > MyIntDictDict;
        this.MyIntDictArrDict = null; //public SerializableDictionary<string, SerializableDictionary<string, int>[]> MyIntDictArrDict;

        this.MyTime = null; // DateTime
        this.MySpan = null; // TimeSpan
    }
    static GetXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        //temp.addString('SomeAttr').attr();
        temp.add('MyEnumProp', 'MyEnum');
        temp.add('MyEnumArray', 'MyEnum', 1);
        temp.addBool('MyBool');
        temp.addByte('MyByte');
        temp.addInt('MyInt');
        temp.addUInt('MyUInt');
        temp.addLong('MyLong');
        temp.addDouble('MyDouble');
        temp.addInt('MyNullInt').nullable();
        temp.addInt('MyNullIntArr',1).nullable();
        temp.addInt('MyIntArray',1);
        //temp.addInt('MyJagIntArray', 2); // jagged array has two dimensions
        temp.addInt('MyJagIntArray', 2, xml.xmlNS_Array); // enum usually doesn't have namespace
        temp.addInt('MyJagIntArray2', 3, xml.xmlNS_Array);
        temp.add('MyJagEnumArray', 'MyEnum',2,"http://schemas.datacontract.org/2004/07/csharpxml.Test1");
        temp.addString('MyStrList', 1);
        temp.add('MySubClass', 'SubTestClass');
        temp.add('MySubClassArray', 'SubTestClass',1);
        temp.add('MyIntDict', 'ArrayOfKeyValueOfstringint');
        temp.add('MyIntDictArr', 'ArrayOfKeyValueOfstringint',1);
        temp.add('MySubClassDict', 'ArrayOfKeyValueOfstringSubTestClassjHm_ShSx9');
        // complex embedding?
        temp.add('MyIntDictDict', 'ArrayOfKeyValueOfstringArrayOfKeyValueOfstringintty7Ep6D1');
        temp.add('MyEnumArrDict', 'ArrayOfKeyValueOfstringArrayOfMyEnumjHm_ShSx9');
        temp.add('MyIntDictArrDict', 'ArrayOfArrayOfKeyValueOfstringint');
        temp.addDateTime('MyTime');
        temp.addTimeSpan('MySpan');
        temp.sortByName(); // cheat to avoid re-ordering these properly (won't be an option with derived classes)
        return temp;
    }
}
module.exports.TestClass=TestClass;

class SubTestClass
{
    constructor()
    {
        //this.SubAttr=0; // [XmlAttribute] public int SubAttr;
        //this.SubAttr2=null; // [XmlAttribute] public string SubAttr2;
        this.MySubInt=0; //public int MySubInt;
        this.MySubStr=null; //public string MySubStr;
    }
    static GetXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        //temp.addInt('SubAttr').attr();
        //temp.addString('SubAttr2').attr();
        temp.addInt('MySubInt');
        temp.addString('MySubStr');
        return temp;
    }
}
module.exports.SubTestClass=SubTestClass;

const MyEnumSimple = { zero:0, one:1, two:2, three:3 }
